### DISCUSSION TITLE
NestJS Microservices - Microservices Decorators Seems to Be Causing Context Action Violation

### DISCUSSION THREAD

**Discussion by EvilCheetah** [mikro-orm/mikro-orm#4256] [Q&A] [answered]
Currently working on the side project that involves NestJS with Microservices in combination with MikroORM.
I have tried to follow the recipe from the official documentation to use EntityRepository approach.
Everything worked fine, until the point I started transforming the project into microservices.

I have figured out empirically that `MessagePattern` method decorator from `@nestjs/microservices` causing **ValidationError** with following message:
`Using global EntityManager instance methods for context specific actions is disallowed. If you need to work with the global instance's identity map, use 'allowGlobalContext' configuration option or 'fork()' instead.`

I have tried to resolve this issue on my own, and have found several discussions regarding the error message above, but none of them seems to be addressing my issue. Moreover, I have found out that turning on `globalContext` without additional knowledge of the framework is HIGHLY UNRECOMMENDED.

I have successfully transformed my controller to use common http decorators, and everything started working as desired, but that's not what I want. What should I do at this point?

**Comment by EvilCheetah** [comment]
Here is a current repository with existing error:

https://github.com/EvilCheetah/coin-catalog

**Reply by B4nan** [reply]
Reproductions need to be minimal and clear, try to simplify it so it contains only of very few files, drop everything unrelated, maybe then you will realize what is happening. NestJS decorators won't trigger this validation on their own, it has to be something they trigger. You should see that from the stack trace.

**Reply by EvilCheetah** [reply]
Hey there!

I have tried to condense everything to two simple services.
Here is a repo: https://github.com/EvilCheetah/mikro-orm-context-error

Everything is hardcoded, so the process to make it work is simple:
- Install dependencies
- Create necessary containers using `docker-compose`
- Migrate the 1 entity using `mikro-orm migration:up`
- Run two applications using `pnpm run start:dev rabbitmq-test` and `pnpm run start:dev gateway`
- Make GET Request to `localhost:3000/`
- Inspect `rabbitmq-test` console
___
When I change `rabbitmq-test` to regular application, by using `Get()` decorator and `create` in main, everything begins to work normally. With such configuration, it keeps throwing the same error.

**Reply by B4nan** [reply, marked as answer]
I see, that sounds like there are no middlewares executed, you would have to handle context creation yourself in such case, e.g. via `em.fork` or the `@UseRequestContext` decorator.

**Reply by EvilCheetah** [reply]
Just out of curiosity, why in one case, when using HTTP approach, it occurs automatically, while in other, in microservice, you have to do it manually?

Is there any way to perform such actions under the hood? Should I investigate into the issue and create a pull request?

Thank you very much for your help and time!

**Reply by B4nan** [reply]
Because http approach supports middlewares, and the orm adapter is registering one automatically to handle this.
